### PR TITLE
Fix differentiation of nested functions inside `@inlinable` functions.

### PR DESCRIPTION
**Pull request by dan-zheng**
Fix differentiation of `@inlinable public func differentiationFunction(from:)`.
Todo: add a test.

### CODE REVIEW

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 1]
```diff
@@ -114,7 +114,8 @@ static SILLinkage getAutoDiffFunctionLinkage(SILLinkage originalLinkage,
   // associated function. Make the associated function public unless
   // differentiation is not explicitly requested.
   if (originalLinkage == SILLinkage::Public ||
-      originalLinkage == SILLinkage::PublicNonABI)
+      originalLinkage == SILLinkage::PublicNonABI ||
+      originalLinkage == SILLinkage::Shared)
```
Comment (dan-zheng) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: In the example of `differentiableFunction(from:)`, the original function is:

```
// original #1 <A, B>(_:) in differentiableFunction<A, B>(from:)
sil shared [serializable] [differentiable source 0 wrt 0 vjp @$s3dif22differentiableFunction4fromq_xcq_5value_15CotangentVectorQzAEQy_c8pullbacktxc_ts14DifferentiableRzsAIR_r0_lF10derivativeL_yq_AD_AfGcAHtxsAIRzsAIR_r0_lF] [ossa] @$s3dif22differentiableFunction4fromq_xcq_5value_15CotangentVectorQzAEQy_c8pullbacktxc_ts14DifferentiableRzsAIR_r0_lF8originalL_yq_xsAIRzsAIR_r0_lF : $@convention(thin) <T, R where T : Differentiable, R : Differentiable> (@in_guaranteed T, @guaranteed @callee_guaranteed (@in_guaranteed T) -> (@out R, @owned @callee_guaranteed (@in_guaranteed R.CotangentVector) -> @out T.CotangentVector)) -> @out R
```

A more precise condition here might be `originalLinkage == SILLinkage::Shared && original->isSerialized()`.
`original->isSerialized()` returns true when `original` is `serialized` or `serializable`.

**Comment by dan-zheng** [comment]
Todo: add a SIL test to `test/AutoDiff`. (Verify that generated JVP/VJPs of original function nested within an `@inlinable` outer function are `shared`, not `hidden`.)

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow linux CPU

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow linux CPU

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow linux CPU

File: lib/SILOptimizer/Mandatory/Differentiation.cpp (hunk 1 above)
Comment (rxwei) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Seems like the right fix. I think that `PublicNonABI` can even be removed here, but let's figure that out later.

**Review by rxwei — APPROVED** [review]

**Pull request merged**